quickwit/quickwit-indexing/src/actors/merge_scheduler_service.rs

```diff
@@ -0,0 +1,463 @@
+use std::sync::Arc;
+
+use anyhow::Context;
+use async_trait::async_trait;
+use quickwit_actors::{Actor, ActorContext, ActorExitStatus, Handler, Mailbox};
+use tantivy::TrackedObject;
+use tokio::sync::{OwnedSemaphorePermit, Semaphore};
+use tracing::error;
+
+use super::MergeSplitDownloader;
+use crate::merge_policy::{MergeOperation, MergeTask};
+
+pub struct MergePermit {
+    _semaphore_permit: Option<OwnedSemaphorePermit>,
+    merge_scheduler_mailbox: Option<Mailbox<MergeSchedulerService>>,
+}
+
+impl MergePermit {
+    #[cfg(test)]
+    pub fn for_test() -> MergePermit {
+        MergePermit {
+            _semaphore_permit: None,
+            merge_scheduler_mailbox: None,
+        }
+    }
+}
+
+impl Drop for MergePermit {
+    fn drop(&mut self) {
+        let Some(merge_scheduler_mailbox) = self.merge_scheduler_mailbox.take() else {
+            return;
+        };
+        if merge_scheduler_mailbox
+            .send_message_with_high_priority(PermitReleased)
+            .is_err()
+        {
+            error!("schedule merge service is dead");
+        }
+    }
+}
+
+pub async fn schedule_merge(
+    merge_scheduler_service: &Mailbox<MergeSchedulerService>,
+    merge_operation: TrackedObject<MergeOperation>,
+    merge_split_downloader_mailbox: Mailbox<MergeSplitDownloader>,
+) -> anyhow::Result<()> {
+    let schedule_merge = ScheduleMerge::new(merge_operation, merge_split_downloader_mailbox);
+    // TODO add backpressure.
+    merge_scheduler_service
+        .ask(schedule_merge)
+        .await
+        .context("failed to acquire permit")?;
+    Ok(())
+}
+
+struct ScheduledMerge {
+    score: u64,
+    id: u64, //< just for total ordering.
+    merge_operation: TrackedObject<MergeOperation>,
+    split_downloader_mailbox: Mailbox<MergeSplitDownloader>,
+}
+
+impl ScheduledMerge {
+    fn order_key(&self) -> (u64, Reverse<u64>) {
+        (self.score, std::cmp::Reverse(self.id))
+    }
+}
+
+impl Eq for ScheduledMerge {}
+
+impl PartialEq for ScheduledMerge {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other).is_eq()
+    }
+}
+
+impl PartialOrd for ScheduledMerge {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for ScheduledMerge {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.order_key().cmp(&other.order_key())
+    }
+}
+
+/// The merge scheduler service is in charge of keeping track of all scheduled merge operation,
+/// and schedule them in the best possible order, and respecting the `merge_concurrency` limit.
+///
+/// This actor is not supervised, and should stay as simple as possible.
+/// In particular,
+/// - the `ScheduleMerge` handler should reply in microseconds.
+/// - the task should never be dropped before reaching its `split_downloader_mailbox` destination
+/// as it would break the consistency of `MergePlanner` with the metastore (ie: several splits will
+/// never) be merged.
```

Comment:
```suggestion
/// The merge scheduler service is in charge of keeping track of all scheduled merge operations,
/// and schedule them in the best possible order, respecting the `merge_concurrency` limit.
///
/// This actor is not supervised and should stay as simple as possible.
/// In particular,
/// - the `ScheduleMerge` handler should reply in microseconds.
/// - the task should never be dropped before reaching its `split_downloader_mailbox` destination
/// as it would break the consistency of `MergePlanner` with the metastore (ie: several splits will
/// never be merged).
```